plenum/test/node_catchup/test_process_catchup_replies.py

```diff
@@ -0,0 +1,156 @@
+    assert set(getattr(reply, f.TXNS.nm).keys()).issubset(received_replies)
+
+
+def check_replies_applied(old_ledger_size, ledger, ledger_info, replies):
+    new_txn_count = sum([len(getattr(reply, f.TXNS.nm).keys())
+                         for reply in replies])
+    assert ledger.size == old_ledger_size + new_txn_count
+    assert ledger.seqNo == old_ledger_size + new_txn_count
+    received_replies = {str(seq_no) for seq_no, _ in ledger_info.receivedCatchUpReplies}
+    assert all(not set(getattr(reply, f.TXNS.nm).keys()).issubset(received_replies)
+               for reply in replies)
+
+
+def test_process_catchup_replies(txnPoolNodeSet, looper, sdk_wallet_client):
+    '''
+    Test correct work of method processCatchupRep and that sending replies
+    in reverse order will call a few iterations of cycle in _processCatchupRep
+    '''
+    # count of transactions and replies (one txn in one reply)
+    # and iterations in _processCatchupRep
+    txns_count = 3
+    ledger_manager = txnPoolNodeSet[0].ledgerManager
+    ledger_info = ledger_manager.getLedgerInfoByType(ledger_id)
+    ledger = ledger_manager.ledgerRegistry[ledger_id].ledger
+    old_ledger_size = ledger.size
+    num_txns_in_reply = 3
+    reply_count = 6
+
+    cons_proof, catchup_reps = _add_txns_to_ledger(txnPoolNodeSet[1],
+                                                   looper,
+                                                   sdk_wallet_client,
+                                                   num_txns_in_reply,
+                                                   reply_count)
+    ledger_info.catchUpTill = cons_proof
+    ledger_info.state = LedgerState.syncing
+
+    # send replies in next order: 2, 3, 1
+    # and check that after sending reply 1, replies 2 and 3 will be applied.
+    reply2 = catchup_reps[1]
+    ledger_manager.processCatchupRep(reply2, sdk_wallet_client[1])
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply2)
+
+    reply3 = catchup_reps[2]
+    ledger_manager.processCatchupRep(reply3, sdk_wallet_client[1])
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply2)
+
+    reply1 = catchup_reps[0]
+    ledger_manager.processCatchupRep(reply1, sdk_wallet_client[1])
+    check_replies_applied(old_ledger_size,
+                          ledger,
+                          ledger_info,
+                          [reply1, reply2, reply3])
+    old_ledger_size = ledger.size
+
+    # send replies in next order: 6, 4, 5
+    # and check that after sending reply 4, it will be applied.
+    # Check that after sending reply 5, replies 5 and 6 will be applied.
+    reply6 = catchup_reps[5]
+    ledger_manager.processCatchupRep(reply6, sdk_wallet_client[1])
+    check_reply_not_applied(old_ledger_size, ledger, ledger_info, reply6)
+
+    reply4 = catchup_reps[3]
+    ledger_manager.processCatchupRep(reply4, sdk_wallet_client[1])
+    check_replies_applied(old_ledger_size, ledger, ledger_info, [reply4])
```

Comment:
Here we should also verify that `reply6` is still not applied.

Reply:
Done